egs/wsj/s5/steps/lmrescore_const_arpa_undeterminized.sh

```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
+# Copyright 2014  Guoguo Chen
+#           2017  Vimal Manohar
+# Apache 2.0
+
+# This script rescores non-compact undeterminized lattices with the 
+# ConstArpaLm format language model.
+# This is similar to steps/lmrescore_const_arpa.sh, but expects 
+# undeterminized non-compact lattices as input.
+
+# Begin configuration section.
+cmd=run.pl
+skip_scoring=false
+stage=1
+scoring_opts=
+write_compact=true
+acwt=0.1
+beam=8.0  # beam used in determinization
+
+# End configuration section.
+
+echo "$0 $@"  # Print the command line for logging
+
+. ./utils/parse_options.sh
+
+if [ $# != 5 ]; then
+  cat <<EOF
+   Does language model rescoring of non-compact undeterminized lattices 
+   (remove old LM, add new LM). This script expects the input lattices 
+   to be in non-compact format.
+   Usage: $0 [options] <old-lang-dir> <new-lang-dir> \\
+                      <data-dir> <input-decode-dir> <output-decode-dir>
+   options: [--cmd (run.pl|queue.pl [queue opts])]
+   See also: steps/lmrescore_const_arpa.sh 
+EOF
+   exit 1;
+fi
+
+[ -f path.sh ] && . ./path.sh;
+
+oldlang=$1
+newlang=$2
+data=$3
+indir=$4
+outdir=$5
+
+oldlm=$oldlang/G.fst
+newlm=$newlang/G.carpa
+! cmp $oldlang/words.txt $newlang/words.txt &&\
+  echo "$0: Warning: vocabularies may be incompatible."
+[ ! -f $oldlm ] && echo "$0: Missing file $oldlm" && exit 1;
+[ ! -f $newlm ] && echo "$0: Missing file $newlm" && exit 1;
+! ls $indir/lat.*.gz >/dev/null &&\
+  echo "$0: No lattices input directory $indir" && exit 1;
+
+if ! cmp -s $oldlang/words.txt $newlang/words.txt; then
+  echo "$0: $oldlang/words.txt and $newlang/words.txt differ: make sure you know what you are doing.";
+fi
+
+oldlmcommand="fstproject --project_output=true $oldlm |"
+
+mkdir -p $outdir/log
+nj=`cat $indir/num_jobs` || exit 1;
+cp $indir/num_jobs $outdir
+
+lats_rspecifier="ark:gunzip -c $indir/lat.JOB.gz |"
+  
+lats_wspecifier="ark:| gzip -c > $outdir/lat.JOB.gz" 
+
+if [ $stage -le 1 ]; then
+  $cmd JOB=1:$nj $outdir/log/rescorelm.JOB.log \
+    lattice-determinize-pruned --acoustic-scale=$acwt --beam=$beam \
```

Comment:
If you are actually running this scripts in your best setups, then it makes me question why you are writing non-determinized lattices at all, because this script determinizes.  If your best results are with rescored lattices, then I think you could substantially simplify things by writing determinized lattice (or maybe just with phone-level determinization).  This script does word-level determinization, though.

Reply:
It composes with the original lattice using lattice-compose at the end. So the final lattice is also not determinized and not word-determinized if the original lattice is not.